test/integration/spokecluster_joining_test.go

```diff
@@ -0,0 +1,113 @@
+package integration_test
+
+import (
+	"context"
+	"path"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
+	clusterv1 "github.com/open-cluster-management/api/cluster/v1"
+	"github.com/open-cluster-management/registration/pkg/helpers"
+	"github.com/open-cluster-management/registration/pkg/spoke"
+	"github.com/open-cluster-management/registration/test/integration/util"
+	"github.com/openshift/library-go/pkg/controller/controllercmd"
+)
+
+var _ = ginkgo.Describe("Joining Process", func() {
+	ginkgo.It("spokecluster should join successfully", func() {
+		var err error
+
+		spokeClusterName := "joiningtest-spokecluster"
+		hubKubeconfigSecret := "joiningtest-hub-kubeconfig-secret"
+		hubKubeconfigDir := path.Join(util.TestDir, "joiningtest", "hub-kubeconfig")
+
+		// run registration agent
+		go func() {
+			agentOptions := spoke.SpokeAgentOptions{
+				ClusterName:         spokeClusterName,
+				BootstrapKubeconfig: bootstrapKubeConfigFile,
+				HubKubeconfigSecret: hubKubeconfigSecret,
+				HubKubeconfigDir:    hubKubeconfigDir,
+			}
+			err := agentOptions.RunSpokeAgent(context.Background(), &controllercmd.ControllerContext{
+				KubeConfig:    spokeCfg,
+				EventRecorder: util.NewIntegrationTestEventRecorder("joiningtest"),
+			})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}()
+
+		// the spoke cluster and csr should be created after bootstrap
+		gomega.Eventually(func() bool {
```

Comment:
I would expect to have to approve the CSR before seeing the spoke cluster come into existence; what am I missing about this?

Reply:
the bootstrap process you expect is the initial implement, the current bootstrap process is after agent is started, it will start two controllers, one creates spoke cluster on hub to wait hub cluster admin to accept and the other creates CSR on hub to wait hub cluster admin to approve it, so the SpokeCluster and CSR will be created almost simultaneously

Reply:
But doesn't the CSR approval gate the initial creation of the SpokeCluster?

Reply:
> But doesn't the CSR approval gate the initial creation of the SpokeCluster?

I haven't looked at the details of the test, but I think that the bootstrap credentials (these are not specific per-cluster, but instead are shared) used by the spokecluster registration controller will have sufficient privilege to create the spokecluster resource.  CSR approval is for credentials specific to a spoke cluster that have a different permissions.  For instance, the specific spoke credentials can read the spoke's namespace from hub and can update status for one particular spoke cluster resource.

Reply:
> I think that the bootstrap credentials (these are not specific per-cluster, but instead are shared) used by the spokecluster registration controller will have sufficient privilege to create the spokecluster resource.

They should have the correct level of permission, but the controller isn't injected with the coordinates of a valid kubeconfig for the hub, so I would expect that the control loop would wait for the CSR to be approved.

I don't want this question to block merge, but I'm failing to see how this can work as-is, so perhaps I'm missing something.

Reply:
I somehow totally missed where this happens in the code. My bad.

Reply:
yes, we still have a loop to wait for CSR to be approved, :)